src/Sarif/Baseline/ResultMatching/DataStructures/MatchedResults.cs

```diff
@@ -173,6 +175,59 @@ private Result CreateBaselinedResult(BaselineState newBaselineState)
             return result;
         }
 
+        internal void SetFirstDetectionTime(Result targetResult)
+        {
+            DateTime firstDetectionTime = default;
+
+            // For existing results (or results that have disappeared), get the first detection
+            // time from the previous result and its associated run. For new results, get the first
+            // detection time from the current result and its associated run.
+            ExtractedResult extractedResult;
+            switch (targetResult.BaselineState)
+            {
+                case BaselineState.Absent:
+                case BaselineState.Updated:
+                case BaselineState.Unchanged:
+                    extractedResult = PreviousResult;
+                    break;
+
+                case BaselineState.New:
+                    extractedResult = CurrentResult;
+                    break;
+
+                default:
+                    throw new ArgumentException("Baseline state was not set.", nameof(targetResult));
+            }
+
+            Result sourceResult = extractedResult.Result;
+            if (sourceResult.Provenance != null && sourceResult.Provenance.FirstDetectionTimeUtc != default)
+            {
+                firstDetectionTime = sourceResult.Provenance.FirstDetectionTimeUtc;
+            }
+            else
+            {
+                Invocation invocation = extractedResult.OriginalRun?.Invocations?[0];
+                if (invocation != null)
+                {
+                    firstDetectionTime = invocation.EndTimeUtc != default
+                        ? invocation.EndTimeUtc
+                        : invocation.StartTimeUtc;
+                }
+            }
+
+            if (firstDetectionTime == default)
+            {
+                firstDetectionTime = DateTime.UtcNow;
+            }
+
+            if (targetResult.Provenance == null)
+            {
+                targetResult.Provenance = new ResultProvenance();
+            }
+
+            targetResult.Provenance.FirstDetectionTimeUtc = firstDetectionTime;
```

Comment:
Should you also set LastDetectionTime to the current Run time for all non-Absent results?